### ISSUE TITLE
[GCP[ Default PodDefaults - add-gcp-secret is invalid

### ISSUE THREAD

**Issue by jlewi** [kubeflow/kubeflow#3924]
/kind bug

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]

If you deploy with kfctl default on platform GCP you end up with a default namespace kubeflow-${USER} that has PodDefaults like the following

```
kubectl -n kubeflow-jlewi get poddefaults add-gcp-secret  -o yaml
apiVersion: kubeflow.org/v1alpha1
desc: add gcp credential
env:
- name: GOOGLE_APPLICATION_CREDENTIALS
  value: /secret/gcp/user-gcp-sa.json
kind: PodDefault
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"kubeflow.org/v1alpha1","env":[{"name":"GOOGLE_APPLICATION_CREDENTIALS","value":"/secret/gcp/user-gcp-sa.json"}],"kind":"PodDefault","metadata":{"annotations":{},"creationTimestamp":"2019-08-16T04:13:51Z","generation":1,"name":"add-gcp-secret","namespace":"kubeflow-jlewi","resourceVersion":"4861","selfLink":"/apis/kubeflow.org/v1alpha1/namespaces/kubeflow-jlewi/poddefaults/add-gcp-secret","uid":"40d741a2-bfdc-11e9-82c7-42010a8e010d"},"spec":{"desc":"add gcp credential","selector":{"matchLabels":{"add-gcp-secret":"true"}}},"volumeMounts":[{"mountPath":"/secret/gcp","name":"secret-volume"}],"volumes":[{"name":"secret-volume","secret":{"secretName":"user-gcp-sa"}}]}
  creationTimestamp: 2019-08-16T04:13:51Z
  generation: 1
  name: add-gcp-secret
  namespace: kubeflow-jlewi
  resourceVersion: "46181"
  selfLink: /apis/kubeflow.org/v1alpha1/namespaces/kubeflow-jlewi/poddefaults/add-gcp-secret
  uid: 40d741a2-bfdc-11e9-82c7-42010a8e010d
spec:
  desc: add gcp credential
  selector:
    matchLabels:
      add-gcp-secret: "true"
volumeMounts:
- mountPath: /secret/gcp
  name: secret-volume
volumes:
- name: secret-volume
  secret:
    secretName: user-gcp-sa

```

The fields volumes, volumeMounts, and description should be in the spec

Before we close this issue

* Get the fix on master
* Get the fix on v0.6-branch
* Open a bug about adding an E2E test

**Comment by jlewi** [comment]
Bug is here:
https://github.com/kubeflow/kubeflow/blob/b0d2e1a68cb2e9fcb874dfbcbae5ac1f560e1366/bootstrap/pkg/kfapp/gcp/gcp.go#L1684

/assign @jlewi

**Comment by jlewi** [comment]
To work around this you can just recreate the poddefault using the following spec

```
apiVersion: kubeflow.org/v1alpha1
kind: PodDefault
metadata:
  name: add-gcp-secret
spec:
  desc: add gcp credential
  env:
  - name: GOOGLE_APPLICATION_CREDENTIALS
    value: /secret/gcp/user-gcp-sa.json
  selector:
    matchLabels:
      add-gcp-secret: "true"
  volumeMounts:
  - mountPath: /secret/gcp
    name: secret-volume
  volumes:
  - name: secret-volume
    secret:
      secretName: user-gcp-sa
```

**Comment by jhua18** [comment]
Hi @jlewi,

I tried the workaround on our KF v0.6 cluster in GCP:
```
## delete current PodDefault
kubectl -n $kubeflow-AN-USER-NAMESPACE delete poddefault add-gcp-secret 
## recreate PodDefault with the yaml
kubectl apply -f add-gcp-secret.yaml -n $kubeflow-AN-USER-NAMESPACE
```
with the add-gcp-secret.yaml file:
```
apiVersion: kubeflow.org/v1alpha1
kind: PodDefault
metadata:
  name: add-gcp-secret
spec:
  desc: add gcp credential
  env:
  - name: GOOGLE_APPLICATION_CREDENTIALS
    value: /secret/gcp/user-gcp-sa.json
  selector:
    matchLabels:
      add-gcp-secret: "true"
  volumeMounts:
  - mountPath: /secret/gcp
    name: secret-volume
  volumes:
  - name: secret-volume
    secret:
      secretName: user-gcp-sa
```
However, [the xgboost_synthetic sample](https://github.com/kubeflow/examples/tree/master/xgboost_synthetic) terminated at the exact same place with same error:
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-1f7852ac284a> in <module>
      3 import os
      4 
----> 5 util.notebook_setup()

~/examples/xgboost_synthetic/util.py in notebook_setup()
     20   subprocess.check_call(["gcloud", "auth", "configure-docker", "--quiet"])
     21   subprocess.check_call(["gcloud", "auth", "activate-service-account",
---> 22                          "--key-file=" + os.getenv("GOOGLE_APPLICATION_CREDENTIALS"),
     23                          "--quiet"])
     24 

TypeError: must be str, not NoneType
```

**original PodDefault before change:**
```
kubectl -n kubeflow-joshua-hua get poddefaults add-gcp-secret -o yaml
apiVersion: kubeflow.org/v1alpha1
desc: add gcp credential
env:
- name: GOOGLE_APPLICATION_CREDENTIALS
  value: /secret/gcp/user-gcp-sa.json
kind: PodDefault
metadata:
  creationTimestamp: 2019-08-13T07:17:36Z
  generation: 1
  name: add-gcp-secret
  namespace: kubeflow-joshua-hua
  resourceVersion: "4717"
  selfLink: /apis/kubeflow.org/v1alpha1/namespaces/kubeflow-joshua-hua/poddefaults/add-gcp-secret
  uid: 6ce322e1-bd9a-11e9-aa77-42010a8a00ef
spec:
  selector:
    matchLabels:
      add-gcp-secret: "true"
volumeMounts:
- mountPath: /secret/gcp
  name: secret-volume
volumes:
- name: secret-volume
  secret:
    secretName: user-gcp-sa
```
**newly created PodDefault:**
```
kubectl -n kubeflow-joshua-hua get poddefaults add-gcp-secret -o yaml
apiVersion: kubeflow.org/v1alpha1
kind: PodDefault
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"kubeflow.org/v1alpha1","kind":"PodDefault","metadata":{"annotations":{},"name":"add-gcp-secret","namespace":"kubeflow-joshua-hua"},"spec":{"desc":"add gcp credential","env":[{"name":"GOOGLE_APPLICATION_CREDENTIALS","value":"/secret/gcp/user-gcp-sa.json"}],"selector":{"matchLabels":{"add-gcp-secret":"true"}},"volumeMounts":[{"mountPath":"/secret/gcp","name":"secret-volume"}],"volumes":[{"name":"secret-volume","secret":{"secretName":"user-gcp-sa"}}]}}
  creationTimestamp: 2019-08-17T05:13:44Z
  generation: 1
  name: add-gcp-secret
  namespace: kubeflow-joshua-hua
  resourceVersion: "1781477"
  selfLink: /apis/kubeflow.org/v1alpha1/namespaces/kubeflow-joshua-hua/poddefaults/add-gcp-secret
  uid: c8ac7d34-c0ad-11e9-aaf6-42010a8a0105
spec:
  desc: add gcp credential
  env:
  - name: GOOGLE_APPLICATION_CREDENTIALS
    value: /secret/gcp/user-gcp-sa.json
  selector:
    matchLabels:
      add-gcp-secret: "true"
  volumeMounts:
  - mountPath: /secret/gcp
    name: secret-volume
  volumes:
  - name: secret-volume
    secret:
      secretName: user-gcp-sa
```

**Comment by jhua18** [comment]
In order to pass the failure point in the sample, a workaround is to add an env variable to point to a service account key json file that must be uploaded to the same location of the notebook first. But, should a SA key file have to be uploaded and added to environment explicitly to complete a notebook setup for an user already has access to the notebook?

**Comment by jlewi** [comment]
@jhua18 In order to apply those PodDefaults to your notebook when you spawn the notebook in the UI you need to select configuration -> and then select "add-gcp-secret".

**Comment by jlewi** [comment]
The fix should now be commited
on v0.6-branch #3926
on master #3934 

As soon as we verify that the fix is working on 0.6 and on master using the auto-deployed clusters we can close this issue.

**Comment by jlewi** [comment]
Last auto-deployed cluster is from 3 days ago.

**Comment by jlewi** [comment]
Testing the latest v0-6-branch via auto-deployments
http://kf-v0-6-n03.endpoints.kubeflow-ci-deployment.cloud.goog/

Pod defaults looks correct.

```
kubectl -n kubeflow-kubeflow-testing  get poddefaults -o yaml add-gcp-secret
apiVersion: kubeflow.org/v1alpha1
kind: PodDefault
metadata:
  creationTimestamp: 2019-08-19T00:19:08Z
  generation: 1
  name: add-gcp-secret
  namespace: kubeflow-kubeflow-testing
  resourceVersion: "4416"
  selfLink: /apis/kubeflow.org/v1alpha1/namespaces/kubeflow-kubeflow-testing/poddefaults/add-gcp-secret
  uid: f62fa30d-c216-11e9-9933-42010a8e0142
spec:
  desc: add gcp credential
  env:
  - name: GOOGLE_APPLICATION_CREDENTIALS
    value: /secret/gcp/user-gcp-sa.json
  selector:
    matchLabels:
      add-gcp-secret: "true"
  volumeMounts:
  - mountPath: /secret/gcp
    name: secret-volume
  volumes:
  - name: secret-volume
    secret:
      secretName: user-gcp-sa
```

**Comment by jhua18** [comment]
@jlewi where is the configuration option in KF jupyter notebook UI?
![Screen Shot 2019-08-19 at 10 31 18 AM](https://user-images.githubusercontent.com/39747424/63286285-85e2ca80-c26c-11e9-864d-e14f543435e1.png)

**Comment by jlewi** [comment]
@jhua18 Its in the configuration to spawn the notebook.
https://github.com/kubeflow/examples/blob/master/xgboost_synthetic/images/addgcpsecret.png

I'm going to close this issue because it should be fixed on the v0.6-branch

**Comment by jhua18** [comment]
@jlewi Got it! That's the configure when spawning a new notebook server. :)
